File: cpp/sockets/multi_client_threaded_server.cpp
#include <winsock2.h>
#include <ws2tcpip.h>
#include <iostream>
#include <stdexcept>
#include <thread>

#pragma comment(lib, "ws2_32.lib")

constexpr const char* PORT = "8888";
constexpr int BUFFER_SIZE = 512;

class WinSock
{
public:
    WinSock()
    {
        if (WSAStartup(MAKEWORD(2, 2), &wsa) != 0)
            throw std::runtime_error("WSAStartup failed");
    }
    ~WinSock()
    {
        WSACleanup();
    }
private:
    WSADATA wsa{};
};

void handle_client(SOCKET client_sock, uint16_t client_port)
{
    char buffer[BUFFER_SIZE];
    int recv_size;

    std::cout << "[+] Client[" << client_port << "] connected\n";

    while ((recv_size = recv(client_sock, buffer, BUFFER_SIZE - 1, 0)) > 0)
    {
        buffer[recv_size] = '\0';
        std::cout << "Client [" << client_port << "]: " << buffer << '\n';

        send(client_sock, buffer, recv_size, 0);
    }

    std::cout << "[-] Client [" << client_port << "] disconnected\n";
    closesocket(client_sock);
}

int main()
{
    try
    {
        WinSock winsock;

        addrinfo hints{}, *res = nullptr;
        hints.ai_family   = AF_INET;
        hints.ai_socktype = SOCK_STREAM;
        hints.ai_flags    = AI_PASSIVE;

        if (getaddrinfo(nullptr, PORT, &hints, &res) != 0)
            throw std::runtime_error("getaddrinfo failed");

        SOCKET listen_sock = socket(res->ai_family, res->ai_socktype, res->ai_protocol);
        if (listen_sock == INVALID_SOCKET)
            throw std::runtime_error("Socket creation failed");

        if (bind(listen_sock, res->ai_addr, (int)res->ai_addrlen) == SOCKET_ERROR)
            throw std::runtime_error("Bind failed");

        freeaddrinfo(res);

        if (listen(listen_sock, SOMAXCONN) == SOCKET_ERROR)
            throw std::runtime_error("Listen failed");

        std::cout << "Server listening on port " << PORT << "...\n";

        while (true)
        {
            sockaddr_in client_addr{};
            int addr_len = sizeof(client_addr);

            SOCKET client_sock = accept(
                listen_sock,
                reinterpret_cast<sockaddr*>(&client_addr),
                &addr_len
            );

            if (client_sock == INVALID_SOCKET)
            {
                std::cerr << "Accept failed\n";
                continue;
            }

            uint16_t client_port = ntohs(client_addr.sin_port);

            std::thread(handle_client, client_sock, client_port).detach();
        }

        closesocket(listen_sock);
    }
    catch (const std::exception& e)
    {
        std::cerr << "Error: " << e.what() << '\n';
        return 1;
    }

    return 0;
}

File: python/sockets/multi_client_threaded_server.py
import socket
import threading

HOST = "0.0.0.0"
PORT = 8888
BUFFER_SIZE = 512

def handle_client(client_sock, addr):
    print(f"[+] Client connected: {addr}")

    with client_sock:
        while True:
            data = client_sock.recv(BUFFER_SIZE)
            if not data:
                break

            message = data.decode()
            print(f"[{addr}] {message}")

            client_sock.sendall(data)

    print(f"[-] Client disconnected: {addr}")

def main():
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as server_sock:
        server_sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        server_sock.bind((HOST, PORT))
        server_sock.listen()

        print(f"Server listening on port {PORT}...")

        while True:
            client_sock, addr = server_sock.accept()

            thread = threading.Thread(
                target=handle_client,
                args=(client_sock, addr),
                daemon=True
            )
            thread.start()

if __name__ == "__main__":
    main()
